### DISCUSSION TITLE
Turbolinks doesn't seem like the best option

### DISCUSSION THREAD

**Discussion by stuartcusackie** [livewire/livewire#1274] [4. General]
Turbolinks causes a few problems that are difficult to solve, and I have decided not to use it in its current state. It would be nice to have an option to remove the Livewire Turbolink code. 

**Turbolink problems:**

- I had to disable Turbolink caching on my Livewire contact form page, otherwise a flicker of text on the form inputs would be seen when switching between pages.
- Any default Laravel error / exception page injects inline CSS, and Turbolinks caches this. Meaning that when you navigate backwards after an exception your design is broken. You would have to publish and customise every Laravel exception page to solve this.
-  It's tricky to make any CSS animations render as normal. Turbolinks switches things out so quickly and doesn't wait for animations to finish. For example, my navbar dropdowns don't look great when a link is clicked.

I am currently experimenting with **Swup.js**, which doesn't cause any of the above problems, and unlike Turbolinks, it comes with some nice page transition effects.

Below is a Swup / AOS.js / Livewire config. I have included AOS to demonstrate how well CSS transitions work with Swup, as opposed to Turbolinks. I also don't have to make any manual fixes to Laravel exception pages with this configuration and my Livewire components work as expected.

I would encourage people to try this out and compare against Turbolinks. I am curious to see if anybody comes to the same conclusion that Swup is much easier to work with than Turbolinks. I will update this post if I come across any Livewire bugs caused by this config. 

Please bear in mind that I am very new to Livewire so I can't say how this will go in the long term but my first Livewire contact form works fine with this config.

```
// AOS first
import AOS from 'aos';
AOS.init({
	startEvent: 'swup:enabled'
})

// Swup second
import Swup from 'swup';
import SwupFadeTheme from '@swup/fade-theme';
const swup = new Swup({
	plugins: [
		new SwupFadeTheme()
	]
});
swup.on('willReplaceContent', unload);
swup.on('contentReplaced', reload);
swup.on('pageLoaded', refreshLivewireComponents)


/*
 * If you need to unload / destroy anything before Swup loads a new page
 */
function unload() {
	// To do....
}

/*
 * Things to run on every Swup reload
 */
function reload() {
	window.livewire.restart()
}

/*
 * Copied from the default Livewire Turbolink script
 */
function refreshLivewireComponents() {

	document.querySelectorAll('[wire\\:id]').forEach(function(el) {
        const component = el.__livewire;

        const dataObject = {
            data: component.data,
            events: component.events,
            children: component.children,
            checksum: component.checksum,
            locale: component.locale,
            name: component.name,
            errorBag: component.errorBag,
            redirectTo: component.redirectTo,
        };

        el.setAttribute('wire:initial-data', JSON.stringify(dataObject));
    });
}
```
You will need to install the packages via npm and set a swup element, e.g.: `<main id="swup">`

**Comment by robertdrakedennis** [comment]
Is there a specific issue you've run into livewire and swup conflicting?

**Reply by stuartcusackie** [reply]
No, none so far.

**Comment by aerni** [comment]
I have been using Swup for a while now. Now the first time in a project together with Livewire. I'm running into issues where Livewire components only work on the initial page load from the server. They stop working as soon as Swup takes over the routing.

I implemented the `reload()` and `refreshLivewireComponents()` function you show above. But it is still not working. Any idea why?

**Reply by aerni** [reply]
I found this plugin for Turbolinks: https://github.com/livewire/turbolinks
This can probably also be ported to work with Swup. Have you had a look at this before?

**Reply by stuartcusackie** [reply]
I'm afraid not! It's been a long time since I wrote this. I remember it took quite a bit of messing around to get it right but I don't think I changed much from the code above.

**Reply by aerni** [reply]
Hmm okay. Hopefully, I can make it work.

**Comment by aerni** [comment]
Alright, I got it working with the following code, which is basically a copy of the [Turbolinks adaptor](https://github.com/livewire/turbolinks):

```js
if (typeof window.Livewire === 'undefined') {
    throw 'Livewire Swup Plugin: window.Livewire is undefined. Make sure @livewireScripts is placed above this script include'
}

let firstTime = true

function wireSwupAfterFirstVisit() {
    // We only want this handler to run AFTER the first load.
    if (firstTime) {
        firstTime = false
        return
    }

    window.Livewire.restart()
}

function wireSwupPageLoaded() {
    document.querySelectorAll('[wire\\:id]').forEach(function (el) {
        const component = el.__livewire
        const dataObject = {
            fingerprint: component.fingerprint,
            serverMemo: component.serverMemo,
            effects: component.effects,
        }
        el.setAttribute('wire:initial-data', JSON.stringify(dataObject))
    })
}

document.addEventListener('swup:pageView', wireSwupAfterFirstVisit)
document.addEventListener('swup:pageLoaded', wireSwupPageLoaded)

```

**Reply by aerni** [reply]
Hmm, actually it still doesn't quite work in some scenarios. I also had to disable the Swup cache. Still figuring things out.

**Reply by stuartcusackie** [reply]
Yes it is a pain alright. Especially when the site needs scripts on some pages but not others. I may recommend to my client not to use Swup just to save myself some pain.

It was easier on my own site which has only a couple of scripts and very few third party scripts. 

By the way, my event sequence differs from what I originally wrote. This is what I use on my own site:

```
swup.on('animationOutStart', resetTransitionClasses)
swup.on('willReplaceContent', unloadDestroyWhatever)
swup.on('contentReplaced', refreshLivewireAndWhateverElse)
```

**Comment by aerni** [comment]
I actually wrote a Swup Plugin to make it work with Livewire: https://github.com/aerni/swup-livewire-plugin

**Reply by stuartcusackie** [reply]
Interesting! I'm having trouble installing with `npm i swup-livewire-plugin`
_Not Found - GET https://registry.npmjs.org/swup-livewire-plugin - Not found_

But I've just included it locally for now.

**Reply by aerni** [reply]
Yeah, I don't think I've published it yet. Sorry about that. I'm talking to the author of Swup to publish it to the official @swup namespace.

**Reply by aerni** [reply]
Let me know how it works for you and please open an issue if you run into anything.

**Reply by stuartcusackie** [reply]
It seems to work the same as my code, which is cool and could save me a lot of messing around in future.

Unfortunately I'm still on the side of not using Swup at all. Livewire behaves strangely with certain properties. I don't know if you have ever seen this error:

**_Livewire encountered corrupt data when trying to hydrate..._**

Which forces me to use protected properties sometimes. This is causing me a problem when using Swup. It seems the swup refresh loses the protected property's value. I'm think I'm giving up trying to solve it!

**Reply by aerni** [reply]
Anyone looking for the solution of this issue check this: https://github.com/aerni/swup-livewire-plugin/issues/1

**Comment by aerni** [comment]
The swup plugin is now officially released. Check it out here: https://swup.js.org/plugins/livewire-plugin

**Reply by stuartcusackie** [reply]
Good stuff! The pre-release version has been working perfectly for me. A very nice addition to Livewire.
